### ISSUE TITLE
More unit tests, especially for oriented or 2d models (Trac #852)

### ISSUE THREAD

**Issue by RichardHeenan** [SasView/sasmodels#141]
From my email of 20 Dec 2016, a survey of unit tests, initially for the "oriented" models, 
and possibly not now completely accurate, shows that we are rather short of unit tests and really ought to add a lot more. 

Fortunately when comparing gpu and cpu computations we can simply use the starting or 
demo parameters, but these may not push the boundaries in quite the same way as a well 
thought unit test.

[After sorting a bug in the new coordinate transform SasView/sasview#931 I later fixed the unit tests in the new coordinates for cylinder (apart from a numerical issue in the 2d test which is sending back a slightly different value) and core_shell_ellipsoid.] 

NOTE that currently the symmetric particles (only theta & phi angles)
are using the new angular coordinates, whilst the asymmetric ones 
(theta, phi & psi angles) are still on the old coordinates, see #123
discussions about angular dispersity.

New angles theta’, phi’ from from old ones are given by:
  cos(theta’) = cos(theta)sin(phi)  
  tan(phi’) = tan(theta)/cos(phi)

Note when writing unit tests, please make sure that all the parameter
values are numerically different, so that if they get shuffled on passing 
through subroutine calls it might be noticed. In the list below "no 2d calc" 
means that angular dispersity is not currently allowed.

'''Cylinder functions ''' 
barbell – no tests
capped_cylinder – no tests
core_shell_bicelle – no tests
core_shell_bicelle_elliptical – one 1d test
core_shell_cylinder – no tests
cylinder – one 1d, one 2d test
elliptical_cylinder – one 1d test
flexible_cylinder – three 1d tests, another at Q=.001 commented out with “currently fails in OCL”, no 2d calc
flexible_cylinder_elliptical – three 1d tests, another at Q=.001 commented out with “currently fails in OCL”, no 2d calc
hollow_cylinder – two 1d tests, 
pearl_necklace – one 1d test, no 2d calc
pringle – three 1d tests, no 2d calc
stacked_disks – all 1d & 2d, tests commented out

'''Ellipsoid Functions'''
core_shell_ellipsoid – two 1d, one 2d test
ellipsoid – no tests
triaxial_ellipsoid – no tests

'''Lamellae Functions'''
lamellar – one 1d test, no 2d calc
lamellar_hg  – one 1d test, no 2d calc
lamellar_hg_stack_caille  – one 1d test, no 2d calc
lamellar_stack_caille  – one 1d test, no 2d calc
lamellar_stack_paracrystal  – two 1d test, no 2d calc
Paracrystal Functions
bcc_paracrystal – no tests, (does 2d calc, has 3 angles)
fcc_paracrystal – no tests, (does 2d calc, has 3 angles)
sc_paracrystal – three 1d tests, (does 2d calc, has 3 angles)
Parallelepiped Functions
core_shell_parallelepiped - two 1d tests, two 2d tests, (has 3 angles, check in new coords?)
hollow_rectangular_prism – one 1d test, no 2d calc
hollow_rectangular_prism_thin_walls – one 1d test, no 2d calc
parallelepiped - one 1d test, two 2d tests, (has 3 angles, check in new coords?)
rectangular_prism – one 1d test, no 2d calc

'''Others'''
I am assuming that all “sphere” and “shape independent” functions are purely 1d.

Migrated from http://trac.sasview.org/ticket/852
```json
{
    "status": "closed",
    "changetime": "2017-04-11T07:18:28",
    "_ts": "2017-04-11 07:18:28.437430+00:00",
    "description": "\nFrom my email of 20 Dec 2016, a survey of unit tests, initially for the \"oriented\" models, \nand possibly not now completely accurate, shows that we are rather short of unit tests and really ought to add a lot more. \n\nFortunately when comparing gpu and cpu computations we can simply use the starting or \ndemo parameters, but these may not push the boundaries in quite the same way as a well \nthought unit test.\n\n[After sorting a bug in the new coordinate transform #836 I later fixed the unit tests in the new coordinates for cylinder (apart from a numerical issue in the 2d test which is sending back a slightly different value) and core_shell_ellipsoid.] \n\nNOTE that currently the symmetric particles (only theta & phi angles)\nare using the new angular coordinates, whilst the asymmetric ones \n(theta, phi & psi angles) are still on the old coordinates, see #776\ndiscussions about angular dispersity.\n\nNew angles theta\u2019, phi\u2019 from from old ones are given by:\n  cos(theta\u2019) = cos(theta)sin(phi)  \n  tan(phi\u2019) = tan(theta)/cos(phi)\n\nNote when writing unit tests, please make sure that all the parameter\nvalues are numerically different, so that if they get shuffled on passing \nthrough subroutine calls it might be noticed. In the list below \"no 2d calc\" \nmeans that angular dispersity is not currently allowed.\n\n'''Cylinder functions ''' \nbarbell \u2013 no tests\ncapped_cylinder \u2013 no tests\ncore_shell_bicelle \u2013 no tests\ncore_shell_bicelle_elliptical \u2013 one 1d test\ncore_shell_cylinder \u2013 no tests\ncylinder \u2013 one 1d, one 2d test\nelliptical_cylinder \u2013 one 1d test\nflexible_cylinder \u2013 three 1d tests, another at Q=.001 commented out with \u201ccurrently fails in OCL\u201d, no 2d calc\nflexible_cylinder_elliptical \u2013 three 1d tests, another at Q=.001 commented out with \u201ccurrently fails in OCL\u201d, no 2d calc\nhollow_cylinder \u2013 two 1d tests, \npearl_necklace \u2013 one 1d test, no 2d calc\npringle \u2013 three 1d tests, no 2d calc\nstacked_disks \u2013 all 1d & 2d, tests commented out\n\n'''Ellipsoid Functions'''\ncore_shell_ellipsoid \u2013 two 1d, one 2d test\nellipsoid \u2013 no tests\ntriaxial_ellipsoid \u2013 no tests\n\n'''Lamellae Functions'''\nlamellar \u2013 one 1d test, no 2d calc\nlamellar_hg  \u2013 one 1d test, no 2d calc\nlamellar_hg_stack_caille  \u2013 one 1d test, no 2d calc\nlamellar_stack_caille  \u2013 one 1d test, no 2d calc\nlamellar_stack_paracrystal  \u2013 two 1d test, no 2d calc\nParacrystal Functions\nbcc_paracrystal \u2013 no tests, (does 2d calc, has 3 angles)\nfcc_paracrystal \u2013 no tests, (does 2d calc, has 3 angles)\nsc_paracrystal \u2013 three 1d tests, (does 2d calc, has 3 angles)\nParallelepiped Functions\ncore_shell_parallelepiped - two 1d tests, two 2d tests, (has 3 angles, check in new coords?)\nhollow_rectangular_prism \u2013 one 1d test, no 2d calc\nhollow_rectangular_prism_thin_walls \u2013 one 1d test, no 2d calc\nparallelepiped - one 1d test, two 2d tests, (has 3 angles, check in new coords?)\nrectangular_prism \u2013 one 1d test, no 2d calc\n\n'''Others'''\nI am assuming that all \u201csphere\u201d and \u201cshape independent\u201d functions are purely 1d.\n\n",
    "reporter": "richardh",
    "cc": "",
    "resolution": "fixed",
    "workpackage": "SasView QA and testing",
    "time": "2017-02-14T15:55:00",
    "component": "sasmodels",
    "summary": "More unit tests, especially for oriented or 2d models",
    "priority": "minor",
    "keywords": "unit tests",
    "milestone": "SasView 4.2.0",
    "owner": "richardh",
    "type": "enhancement"
}
```

**Comment by butlerpd** [comment]
Trac update at `2017/03/21 01:27:52`: **butler** changed workpackage from "SasView Bug Fixing" to "SasView QA and testing"

**Comment by butlerpd** [comment]
Trac update at `2017/03/21 01:28:16`: **butler** changed milestone from "sasmodels Next Release +1" to "SasView 4.2.0"

**Comment by RichardHeenan** [comment]
Trac update at `2017/04/11 07:18:08`: **richardh** commented:

All the models listed above now have 1d and where appropriate 2d unit tests of some sort. Many of these are included in the sasmodels branch for #147.

There likely other models that still need unit tests, but I will close this ticket.

**Comment by RichardHeenan** [comment]
Trac update at `2017/04/11 07:18:28`: 
* **richardh** changed resolution from "" to "fixed"
* **richardh** changed status from "new" to "closed"
